### PR TITLE
[postcli] option to look for lost VRF nonce in initialized data

### LINKED ISSUE #190: Add functionality to look for a VRF nonce in already initialized data

### PR DESCRIPTION
**Pull request by poszu**
Closes #190 

- [ ] need to see if doing it single-threaded is sufficiently fast.

### CODE REVIEW

File: initialization/vrf_search.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		default:
+			// continue looking for a nonce
+		}
+
+		_, err := io.ReadFull(r, labelBuf)
+		switch {
+		case err == io.EOF:
+			return nonce, nonceValue, nil
+		case err == io.ErrUnexpectedEOF:
+			return 0, nil, errors.New("unexpected end of file - file is truncated. please reinit it")
```
Comment (fasmat) [inline initialization/vrf_search.go]:
```suggestion
			return 0, nil, fmt.Errorf("file appears truncated. please re-init it: %w", err)
```

File: initialization/vrf_search.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+		return true, nil
+	} else if comp == 0 {
+		// need to regenerate label to verify lower bytes
+		res, err := wo.Position(index)
+		if err != nil {
+			return false, fmt.Errorf("failed to regenerate label: %w", err)
+		}
+		return bytes.Compare(res.Output, difficulty) < 0, nil
+	}
+	return false, nil
```
Comment (fasmat) [inline initialization/vrf_search.go]: Would it make sense to use switch here?

```suggestion
	switch {
	case comp < 0:
		return true, nil
	case comp == 0:
		// need to regenerate label to verify lower bytes
		res, err := wo.Position(index)
		if err != nil {
			return false, fmt.Errorf("failed to regenerate label: %w", err)
		}
		return bytes.Compare(res.Output, difficulty) < 0, nil
	default: // comp > 0
		return false, nil
	}
```

File: initialization/vrf_search.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+		return nonce, nonceValue, nil
+	}
+	return 0, nil, ErrNonceNotFound
+}
+
+// searchForNonce searches for a nonce in the given reader.
+func searchForNonce(ctx context.Context, r io.Reader, difficulty []byte, oracle *oracle.WorkOracle) (nonce uint64, nonceValue []byte, err error) {
+	labelBuf := make([]byte, postrs.LabelLength)
+	labelIndex := uint64(0)
+	for {
```
Comment (fasmat) [inline initialization/vrf_search.go]:
```suggestion
	labelIndex := uint64(0)
	for labelIndex := uint64(0); ; labelIndex++ {
```

File: initialization/vrf_search.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+			return 0, nil, errors.New("unexpected end of file - file is truncated. please reinit it")
+		}
+
+		ok, err := checkLabel(labelIndex, labelBuf, difficulty, oracle)
+		if err != nil {
+			return 0, nil, fmt.Errorf("checking label: %w", err)
+		}
+		if ok {
+			nonce = labelIndex
+			nonceValue = append(nonceValue[:0], labelBuf...)
```
Comment (fasmat) [inline initialization/vrf_search.go:180]:
```suggestion
			copy(nonceValue, labelBuf)
```

File: initialization/vrf_search.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+			difficulty = nonceValue // override difficulty to the new lowest label
```
Comment (fasmat) [inline initialization/vrf_search.go:181]: Maybe also `copy` just to be safe?

File: initialization/vrf_search.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+		idx, l, err := searchForNonce(ctx, bufio.NewReader(file), difficulty, woReference)
+		if err != nil {
+			return 0, nil, fmt.Errorf("failed to search for nonce: %w", err)
+		}
+		if l != nil {
+			nonceValue = l
+			nonce = firstLabelIndex + idx
+			logger.Info("found nonce", zap.Uint64("index", nonce), zap.String("label", hex.EncodeToString(nonceValue)))
+			difficulty = nonceValue // override difficulty to the new lowest label
```
Comment (fasmat) [inline initialization/vrf_search.go:127]: maybe `copy` just to be safe?

File: initialization/vrf_search.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+	}
+}
+
+// checkLabels checks if label is lower than difficulty.
+// It will regenerate the whole 32B of the label if its most significant 16B are equal to difficulty
+// in order to check the lower bytes.
+// * index:      the index of the label (used to regenerate it)
+// * label:		 16B most significant bytes of the label
+// * difficulty: all 32b of the difficulty
+// * wo:         the oracle used to regenerate the label.
```
Comment (fasmat) [inline initialization/vrf_search.go]: Formatting appears to be off?

File: initialization/vrf_search.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+// searchForNonce searches for a nonce in the given reader.
+func searchForNonce(ctx context.Context, r io.Reader, difficulty []byte, oracle *oracle.WorkOracle) (nonce uint64, nonceValue []byte, err error) {
+	labelBuf := make([]byte, postrs.LabelLength)
+	labelIndex := uint64(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return nonce, labelBuf, ctx.Err()
```
Comment (fasmat) [inline initialization/vrf_search.go]: The way it is implemented at the moment if once ctrl+C the search after a nonce was found it won't be saved. Only after the search is completed.

Intended?

**Comment by fasmat** [comment]
> * [ ]  need to see if doing it single-threaded is sufficiently fast.

I'm 99% sure the bottleneck is the read speed from disk in most cases at least that's how we designed PoST 🙂

**Comment by poszu** [comment]
> > * [ ]   need to see if doing it single-threaded is sufficiently fast.
> 
> I'm 99% sure the bottleneck is the read speed from disk in most cases at least that's how we designed PoST 🙂

Yes, I also think so.

File: initialization/vrf_search.go (hunk 4 above)
Comment (poszu) [reply initialization/vrf_search.go:180]: I use append to copy because initially `nonceValue` is `nil`.

File: initialization/vrf_search.go (hunk 5 above)
Comment (poszu) [reply initialization/vrf_search.go:181]: Do you mean to not modify `difficulty` by mistake by modifying nonceValue?

File: initialization/vrf_search.go (hunk 8 above)
Comment (poszu) [reply initialization/vrf_search.go]: Thanks, good catch :+1:

**Comment by poszu** [comment]
@pigmej would you like to review the documentation?

**Review by pigmej — APPROVED** [review]
Good docs addition :)

File: initialization/vrf_search.go (hunk 5 above)
Comment (fasmat) [reply initialization/vrf_search.go:181]: yes. Just as a safety measure, in terms of performance it should make little to no difference.

**Review by fasmat — APPROVED** [review]

**Pull request merged**